Layer effect

plasticviking opened the pull request:
# Overview

This PR includes the following proposed change(s):

- {List all the changes, if possible add the linked issue/ticket #}

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots if applicable.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes

plasticviking committed:
wip

plasticviking committed:
fix source cleanup and layer stacking

plasticviking committed:
wip

plasticviking committed:
Merge branch 'dev' into layer-effect

# Conflicts:
#	app/src/UI/Features/LegacyMap/helpers/components/Coordinates/Coordinates.tsx
#	app/src/UI/Features/LegacyMap/helpers/functional/recordset-layers.ts

plasticviking committed:
finish up wms layers, disable sqlite on android

plasticviking committed:
remove reactive-layers

LocalNewsTV commented on app/src/UI/Features/LegacyMap/Controls/PrimaryLayerSelect.tsx:
```diff
@@ -1,21 +1,20 @@
-  const { baseMapLayer, availableBaseMapLayers } = useSelector((state) => state.Map);
-  const offlineDefinitions = useSelector((state) => state.TileCache?.mapSpecifications) ?? [];
+const PrimaryLayerSelect = ({ layers, selectLayer }: PrimaryLayerSelectProps) => {
+  // const offlineDefinitions = useSelector((state) => state.TileCache?.mapSpecifications) ?? [];
```
crumbs

LocalNewsTV commented on app/src/UI/Features/LegacyMap/LayerPicker/LpLayers/LpLayers.tsx:
```diff
@@ -14,10 +16,16 @@ const EmptyCollection = ({ text }: EmptyListProps) => (
+};
+
+const LpLayers = ({ layers, setOverlayState }: PropTypes) => {
+  // const handleWmsClick = (layer) => {
+  //   dispatch({ type: TOGGLE_WMS_LAYER, payload: { layer } });
+  // };
```
crumbs

LocalNewsTV commented on app/src/UI/Features/LegacyMap/LayerPicker/LayerPicker.tsx:
```diff
@@ -88,20 +95,20 @@ export const LayerPicker = () => {
                       <hr />
                     </li>
-                    {MOBILE && (
+                    <FeatureGated requires={'CACHE_TILES'}>
```
One was changed to a `CACHE_TILES` feature gate, the other is still `MOBILE`, so two different conditions depending on LP State

LocalNewsTV commented on app/src/UI/Features/LegacyMap/helpers/functional/layers-hook.ts:
```diff
@@ -0,0 +1,195 @@
+    const updatedLayerDefinitions: InvasivesMapLayerDefinitionWithState[] = [];
+
+    for (const l of availableLayerDefinitions) {
+      const updated = l;
```
```suggestion
      const updated = { ...l };
```
This seems to correct the issue of immutability. good 'ol memory references

LocalNewsTV commented on app/src/UI/Features/LegacyMap/helpers/functional/layer-definitions/baked-vector-simplified.ts:
<img width="851" height="80" alt="image" src="https://github.com/user-attachments/assets/2876944c-a9fa-46b4-9fbf-29edc16d7bd1" />
Stale bread, or part of one of the branches to be merged in?

LocalNewsTV reviewed (commented):
This immutability error occurs often on load. first load on Mobile it happened 100% of the time. Could circumvent it by loading web mode, loading page, then changing to mobile mode. Performing a hard refresh caused the error to happen again.
<img width="428" height="81" alt="image" src="https://github.com/user-attachments/assets/7ccc6e1d-6812-4798-abe7-80cfe591911f" />

Error results in a blue screen. Tend to notice it more in the web mobile version than the web version, though it happens in both.

**WMS Layers are not working when enabled. On Mobile/Web they result in a 400 error**
<img width="485" height="76" alt="image" src="https://github.com/user-attachments/assets/53f3a9af-1302-4988-b3a3-d89b166fd0f9" />

**On Android, they result in this error**
<img width="974" height="72" alt="image" src="https://github.com/user-attachments/assets/38c80f12-5fc1-4d80-88c9-62fbed29bd86" />

When comparing the request between Dev/Local I notice that the `bbox` key isn't populating the same (left: Local, right: Dev)
<img width="678" height="326" alt="image" src="https://github.com/user-attachments/assets/8c8c3eb4-0d7f-4db2-b3ca-144789a1d0c1" />

WMS layers no longer persist on a page refresh, or when going between online -> Offline. I expect the end users to find the second case especially annoying if they have a brief connection outtage. 

Draw tool dots don't show up on Map, these layer definitions are seemingly not-present.
Draw tools themselves work, the edit functionality works, but all visual indicators are gone. 

- `handle_USER_SETTINGS_GET_INITIAL_STATE_REQUEST` makes a `listRepositories` request that consistently crashes on my physical android device. During testing this only occurred on my physical android device (4GB) and didn't occur on a simulated 8GB, or Web

## TL;DR
- Unable to properly test WMS Layers due to bad requests.
- No visual indicators for draw tools
- Suggestion left that seems to fix immutability error causing blue screen crash.
- App feels more peformant from the state cleanup/flow-control
- Issue with Recordset DB being ready.

plasticviking committed:
Map layer effect

- cleanup stale code
- add simplified vector layer
- persist preferred WMS and basemap layer settings
- switch Android to localforage for cached recordsets
- fix drawcontrols layer stacking
- use immer for state updates

plasticviking committed:
add offline layers

plasticviking committed:
remove extra log

LocalNewsTV reviewed (changes requested):
> [!NOTE]
> Not everything in the Review might be completely related to the PR. But this is the first PR Where the physical Android devices have been able to run the application without crashing, and so other non-related issues might be mentioned.
> Some issues might be mentioned multiple times if they were consistent across devices.

# Code

Nothing stands out as egregious that I can pinpoint/target, but there are issues in the flow control.

# [Web Testing]

On first entry into the app (Cleared App Cache/Private Browser) The base map layer buttons and map attributions rapidly fire between two states until one is selected by the user.
- This only occurs on Web, does not occur on Mobile
- Once a field is selected, the issue doesn’t re-occur due to the persistence

Offline Map Tiles Layer Picker Option available in Web when shouldnt be

Base map option buttons don't appear when offline

<img width="313" height="91" alt="image" src="https://github.com/user-attachments/assets/43d01712-c343-4758-ae59-589517dabe4c" />

A useEffect doesn’t  stop firing, resulting Ian a `maximum update depth exceeded` error.
- This only happens in Web, and not in Mobile

 <img width="594" height="331" alt="image" src="https://github.com/user-attachments/assets/c30f29b2-d211-43e1-a3c6-33b79768eaae" />

This error gets thrown but both components continue to work as intended. You can get this by refreshing on the `/Records` page.
- This issue does not happen when running `Mobile`
- 
<img width="599" height="98" alt="image" src="https://github.com/user-attachments/assets/7a98ec99-7542-4e83-bb81-5caa7f2ef41a" />

WMS Layers

- Persist on refresh and toggling Offline/Online.
- WMS Layers are working as expected
- When Offline, the Unavailable text no longer renders 

<img width="358" height="145" alt="image" src="https://github.com/user-attachments/assets/18c05ea1-0d4c-4c5a-b849-2eab955a25b3" />

Base Maps

- BaseMaps are working as 
- BaseMaps choice persists between loads

Draw Tools
- Now behave and show as expected


# [iOS Testing]

WMS Layers
- WMS Layers not working, requests come back like this
<img width="1183" height="123" alt="image" src="https://github.com/user-attachments/assets/8e9c4ace-b0da-43dc-a585-bde74310b685" />
- Notice this error a lot, not sure if its consequences  
<img width="974" height="79" alt="image" src="https://github.com/user-attachments/assets/5ca533c6-b786-4857-8a0c-68b6e1b59cc5" />

Base Maps
-  Base Map Options don't appear when offline. 
    - Gray maps is the default, does not default into the baked tiles.

Offline Maps
- Can Download Offline Maps
- Can Enable Offline Maps from Layer Picker
    - Cannot enable offline maps from /OfflineMaps
  
# [Android Testing] - Physical Lenovo 4GB Ram Device

>[!IMPORTANT]
> Android didn’t crash a single time during testing :)
>App **did not** crash from going into offline mode and **not** having baked map tiles (Watching the Resource Monitor did not result in a climb of CPU). Previously this was a fast crash.

> Testing exclusively on the Normal layout, not that decoupled one for dev.

Base Maps
- Base Map options not available when offline
WMS Layers
	- WMS Layers not available as option

Overlay Behaviour
- Initially changing the Overlay positioning did not work. Touchees and Drags resulted in a downward drag of roughly 50px
    - The Behaviour feels like its keeping the overlay handler as the active target, so the next tap continues this issue, and the one following behaves as expected
- Oddly, after 15 or so minutes of testing/comparing results, the Overlay began to behave as expected
    - After this happened, the items in the Overlay no longer scrolled. Items visible can be interacted with as expected, the drag event is getting consumed.

Offline Maps
- Was able to download Offline Map tiles on device
- Layer Picker did not show Offline Map Tiles
- Attempting to enable Offline Map Tiles from /OfflineMaps resulted in nothing happening
    - Toggling Recordsets still worked, so this is not an `icon button` issue
- Inconsistency with the Feature Flags around this?

# [Android Testing] - 8GB Simulator

Same Overlay behaviour as before. Didn't get the Overlay to work though, but still not sure how I triggered it the first time.
- Sometimes Dragging the Map will also move the Overlay.

WMS Layers
- WMS Layers Available to select
- WMS Layers do not work when toggled on

BaseTile Options not available when Offline

Offline Maps
- Can Download Offline Maps
- Can Enable Offline Maps from Layer Picker
- Cannot Enable Offline Maps from /OfflineMaps page

## [Majors Summary]
- When Offline, BaseMap options are not available, and it defaults into gray maps. This makes the offline functionality unusable in current form
- Offline Map Tiles are being initialized in the /OfflineMaps section, so the Cached Map Tiles area of the Layer Picker Doesn't get populated until you've been on the /OfflineMaps page.


## [Concerns]
This branch is not in sync with dev currently. I am unable to test if the download continuity of Offline Maps is still in working order. 

## [Other Notes]

It seems the Layer Picker is firing a different action than the Offline Maps page, and is the likely culpriot
LP - `togglePreferredOverlayLayer`
Maps Page - `toggleOverlay`

<img width="354" height="133" alt="image" src="https://github.com/user-attachments/assets/77220656-ddec-4eb4-9959-1fa6908048da" />

plasticviking committed:
Merge branch 'dev' into layer-effect

# Conflicts:
#	app/src/UI/Features/LegacyMap/Controls/ButtonContainer.tsx

plasticviking committed:
fix feature flags, wms cache, memoization race

LocalNewsTV reviewed (commented):
## iOS

- Baked tiles non-selectable when offline. This would make app undeployable.
- [Fixed] WMS Layers are now working
- [Fixed] Caching / Displaying working
-  Map tile continuity appears to work as expected

## Android - 4gb

- Still not crashing
- WMS doesn't display with a weird render of no-table, because it doesnt render at all
- Drag issue still present but not part of this branch
- Can download offline maps but not use them. Neglible since we don't have a full release
- Baked Maps don't work

## Android 8gb Sim

- [?] Unsure if WMS Layers work due to API changes and the simulator won't take my ENV changes to point locally
- Can Download/enable maps
- Also doesn't crash

## Web
- [Fixed] Tile Picker availability removed 
- [Fixed] useEffect chain no longer firing


## Bonus
- [Fixed] Toggling off Offline Map caches works now. This is an existing problem in dev

plasticviking committed:
fix offline layers on ios

plasticviking committed:
ensure at least one basemap is always selected
